### PR TITLE
hardening/629_add_complex_configuration_examples

### PR DESCRIPTION
**Pull request by frbattid**
Implements issue #629

### CODE REVIEW

File: doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md [hunk 1]
```diff
@@ -1,47 +1,470 @@
...
+
+Analogous to the FIWARE service, the FIWARE service path can be used to parallelize the processing of the context data notified by Orion Context Broker to Cygnus. Either by creating one listener per service-service path pair (then a Cygnus agent can be created per listener, or a single Cygnus agent with multiple listeners can be created), either creating a single listener for all the service-service path pairs and internarlly to Cygnus multiplex the data based on the FIWARE service-path header.
+
+Regarding Orion Context Broker subscriptions, if multiple listeners are used (either multiple agents either a single agent with many listeners) a subscription for each service-service path pair is required, pointing each one to a different endpoint URL. Nevertheless, if a single Cygnus agent with a single listener, internally multiplexing context data based on the FIWARE service path, is used, then a single suscription can be made for all the service paths within the client FIWARE service. This is achieved by using a special FIWARE service path header's value: `/#`.
```
Comment (fgalan) [inline doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md]: service-service path-> service-servicepath ?

File: doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md [hunk 2]
```diff
@@ -1,47 +1,470 @@
...
 
 [Top](#top)
 
-##<a name="section5"></a>Multiple sources 
-To be done
+### <a name="section2.2"></a>Scenarios involving multiple FIWARE services
+Another header, the **FIWARE service path**, can be used to segmentate data belonging to a single tenant or client. Such a header is used by Cygnus to prefix the name of tables and collections in database-based sinks, or as a subfolder within the HDFS user space, or as a package/dataset in CKAN, etc.
+
+Analogous to the FIWARE service, the FIWARE service path can be used to parallelize the processing of the context data notified by Orion Context Broker to Cygnus. Either by creating one listener per service-service path pair (then a Cygnus agent can be created per listener, or a single Cygnus agent with multiple listeners can be created), either creating a single listener for all the service-service path pairs and internarlly to Cygnus multiplex the data based on the FIWARE service-path header.
```
Comment (fgalan) [inline doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md]: service-service path-> service-servicepath ?

Comment (fgalan) [inline doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md]: internarlly -> internally

File: doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md [hunk 3]
```diff
@@ -1,47 +1,470 @@
...
+#### <a name="section2.1.4"></a>Orion Context Broker subscriptions
+It seems obvious a per client subscription is required when a per client listener is configured for Cygnus (either by using multiple agents, either by using a single agent with multiple listeners). In that case, each subscription will point to a different endpoint URL.
+
+Nevertheless, it may not seem so obvious how subscriptions look like when a single Cygnus agent with a single listener internally multiplexes the notified context data based on the FIWARE service. In that case, a suscription per client is also required (since Orion does not allow to subscribe to all FIWARE services, due to security reasons), but all subscriptions must point to the same single endpoint URL.
 
 [Top](#top)
 
-##<a name="section5"></a>Multiple sources 
-To be done
+### <a name="section2.2"></a>Scenarios involving multiple FIWARE services
```
Comment (fgalan) [inline doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md]: No picture and configuration fragment in this case? Maybe too complex?

File: doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md [hunk 4]
```diff
@@ -1,47 +1,470 @@
...
+
+```
++-----------------------------------+
+|  +-----+     +----+     +------+  |
+|  | src |-----| ch |-----| sink |  |
+|  +-----+     +----+     +------+  |
++-----------------------------------+
+```
+
+Its agent configuration file is the usal one:
```
Comment (frbattid) [inline doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1420/commits/ea511e5c58c65fc69304692013f33e82ae986f1a

File: doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md [hunk 5]
```diff
@@ -1,47 +1,470 @@
...
+cygnus-ngsi.channels.ch2.type = ...
+cygnus-ngsi.channels.ch2...
+
+...
+
+cygnus-ngsi.channels.chN.type = ...
+cygnus-ngsi.channels.chN...
+```
+
+Use multi-backend configuration when you simply want the same NGSI context data to be replicated in each one of the backends. This is possible thanks to the default channel selector, which [replicates](https://flume.apache.org/FlumeUserGuide.html#replicating-channel-selector-default) internal NGSI events to all the channels connected to the source. Such a replication is always done independently of the number of FIWARE service paths you own within your FIWARE service, or the number of kind of entities you have.
```
Comment (frbattid) [inline doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1420/commits/ea511e5c58c65fc69304692013f33e82ae986f1a

File: doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md [hunk 6]
```diff
@@ -1,47 +1,470 @@
...
 
 [Top](#top)
 
-##<a name="section2"></a>Single source, multiple storages
-To be done
+## <a name="section2"></a>Advanced configurations
+### <a name="section2.1"></a>Scenarios involving multiple FIWARE services
+Cygnus NGSI, as many of the FIWARE components, natively supports multitenancy thanks to the **FIWARE service** concept. NGSI notifications consumed by Cygnus contain a Http header carrying such a value, which is used for segmentating context information in the historical backend in the form of per-service dedicated MySQL/MongoDB/STH database, or HDFS user space, or CKAN organization, etc.
```
Comment (frbattid) [inline doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1420/commits/ea511e5c58c65fc69304692013f33e82ae986f1a

File: doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md [hunk 7]
```diff
@@ -1,47 +1,470 @@
...
+cygnus-ngsi-<id>.sinks.sink-client-<id>.type = ...
+cygnus-ngsi-<id>.sinks.sink-client-<id>.channel = ch-client-<id>
+cygnus-ngsi-<id>.sinks.sink-client-<id>...
+
+# channels
+cygnus-ngsi-<id>.channels.ch-client-<id>.type = ...
+cygnus-ngsi-<id>.channels.ch-client-<id>...
+```
+
+[Multi-instance](ref) support is available form the very begining of Cygnus, through the `cygnus_instance_<id>.conf` configuration files used when running Cygnus [as a service](ref):
```
Comment (frbattid) [inline doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1420/commits/ea511e5c58c65fc69304692013f33e82ae986f1a

File: doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md (hunk 3 above)
Comment (frbattid) [inline doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1420/commits/ea511e5c58c65fc69304692013f33e82ae986f1a

File: doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md [hunk 8]
```diff
@@ -1,47 +1,470 @@
...
+It seems obvious a per client subscription is required when a per client listener is configured for Cygnus (either by using multiple agents, either by using a single agent with multiple listeners). In that case, each subscription will point to a different endpoint URL.
+
+Nevertheless, it may not seem so obvious how subscriptions look like when a single Cygnus agent with a single listener internally multiplexes the notified context data based on the FIWARE service. In that case, a suscription per client is also required (since Orion does not allow to subscribe to all FIWARE services, due to security reasons), but all subscriptions must point to the same single endpoint URL.
 
 [Top](#top)
 
-##<a name="section5"></a>Multiple sources 
-To be done
+### <a name="section2.2"></a>Scenarios involving multiple FIWARE services
+Another header, the **FIWARE service path**, can be used to segmentate data belonging to a single tenant or client. Such a header is used by Cygnus to prefix the name of tables and collections in database-based sinks, or as a subfolder within the HDFS user space, or as a package/dataset in CKAN, etc.
```
Comment (frbattid) [inline doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1420/commits/ea511e5c58c65fc69304692013f33e82ae986f1a

File: doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md (hunk 2 above)
Comment (frbattid) [inline doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1420/commits/ea511e5c58c65fc69304692013f33e82ae986f1a

File: doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md (hunk 1 above)
Comment (frbattid) [inline doc/cygnus-ngsi/installation_and_administration_guide/configuration_examples.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1420/commits/ea511e5c58c65fc69304692013f33e82ae986f1a

**Comment by fgalan** [comment]
LGTM

**Pull request merged**